p2p/test/transport/gating_test.go

```diff
@@ -0,0 +1,234 @@
+
+func stripCertHash(addr ma.Multiaddr) ma.Multiaddr {
+	for {
+		if _, err := addr.ValueForProtocol(ma.P_CERTHASH); err != nil {
+			break
+		}
+		addr, _ = ma.SplitLast(addr)
+	}
+	return addr
+}
+
+func TestInterceptPeerDial(t *testing.T) {
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			connGater := NewMockConnectionGater(ctrl)
+
+			h1 := tc.HostGenerator(t, TransportTestCaseOpts{NoListen: true, ConnGater: connGater})
+			h2 := tc.HostGenerator(t, TransportTestCaseOpts{})
+			require.Len(t, h2.Addrs(), 1)
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			connGater.EXPECT().InterceptPeerDial(h2.ID())
+			require.ErrorIs(t, h1.Connect(ctx, peer.AddrInfo{ID: h2.ID(), Addrs: h2.Addrs()}), swarm.ErrGaterDisallowedConnection)
+		})
+	}
+}
+
+func TestInterceptAddrDial(t *testing.T) {
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			connGater := NewMockConnectionGater(ctrl)
+
+			h1 := tc.HostGenerator(t, TransportTestCaseOpts{NoListen: true, ConnGater: connGater})
+			h2 := tc.HostGenerator(t, TransportTestCaseOpts{})
+			require.Len(t, h2.Addrs(), 1)
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			gomock.InOrder(
+				connGater.EXPECT().InterceptPeerDial(h2.ID()).Return(true),
+				connGater.EXPECT().InterceptAddrDial(h2.ID(), h2.Addrs()[0]),
+			)
+			require.ErrorIs(t, h1.Connect(ctx, peer.AddrInfo{ID: h2.ID(), Addrs: h2.Addrs()}), swarm.ErrNoGoodAddresses)
+		})
+	}
+}
+
+func TestInterceptSecuredOutgoing(t *testing.T) {
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			connGater := NewMockConnectionGater(ctrl)
+
+			h1 := tc.HostGenerator(t, TransportTestCaseOpts{NoListen: true, ConnGater: connGater})
+			h2 := tc.HostGenerator(t, TransportTestCaseOpts{})
+			require.Len(t, h2.Addrs(), 1)
+			require.Len(t, h2.Addrs(), 1)
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			gomock.InOrder(
+				connGater.EXPECT().InterceptPeerDial(h2.ID()).Return(true),
+				connGater.EXPECT().InterceptAddrDial(h2.ID(), gomock.Any()).Return(true),
+				connGater.EXPECT().InterceptSecured(network.DirOutbound, h2.ID(), gomock.Any()).Do(func(_ network.Direction, _ peer.ID, addrs network.ConnMultiaddrs) {
+					// remove the certhash component from WebTransport addresses
+					require.Equal(t, stripCertHash(h2.Addrs()[0]).String(), addrs.RemoteMultiaddr().String())
+				}),
+			)
+			require.Error(t, h1.Connect(ctx, peer.AddrInfo{ID: h2.ID(), Addrs: h2.Addrs()}))
```

Comment:
What error are we expecting?

Reply:
Depends on the transport. All we can assert here is that some kind of error occurred.

Reply:
Can we do better? Should we return a well known error? Callers may want to check if this error case happened and it doesn’t make sense for them to check against all error types depending on underlying transport.